### PR TITLE
Shared FFN Weights for the atom/bond targets PR

### PR DESCRIPTION
**Pull request by cjmcgill**
This is a PR related to the atom-bond targets PR currently being developed. This is a change to allow atom FFNs to share weights between tasks and the same for bond FFNs.

I don't have datasets and tests ready to see if it's working yet easily. @shihchengli  maybe you could share the data or scripts that you are using?

Still to do: add an option argument that allows the FFN weights to either be shared or not.

### CODE REVIEW

**Comment by cjmcgill** [comment]
Okay now there is an optional argument controlling whether the weights are shared or not.

File: chemprop/models/ffn.py [hunk 1]
```diff
@@ -42,17 +43,47 @@ def __init__(self,
...
+        else:
+            self.atom_ffn_base = nn.Identity()
+            self.bond_ffn_base = nn.Identity()
+
         ind = 0
+
         for constraint in atom_constraints:
             self.add_module(f'readout_{ind}', FFNAtten(features_size, hidden_size, num_layers, output_size,
-                                                       dropout, activation, constraint, ffn_type='atom'))
+                                                       dropout, activation, self.atom_ffn_base, constraint, ffn_type='atom'))
```
Comment (shihchengli) [inline chemprop/models/ffn.py:81]:
```suggestion
                                                       dropout, activation, self.atom_ffn_base, constraint, ffn_type='atom', shared_ffn))
```

File: chemprop/models/ffn.py [hunk 2]
```diff
@@ -42,17 +43,47 @@ def __init__(self,
...
             ind += 1
 
         for constraint in bond_constraints:
             self.add_module(f'readout_{ind}', FFNAtten(features_size, hidden_size, num_layers, output_size,
-                                                       dropout, activation, constraint, ffn_type='bond'))
+                                                       dropout, activation, self.bond_ffn_base, constraint, ffn_type='bond'))
```
Comment (shihchengli) [inline chemprop/models/ffn.py:86]:
```suggestion
                                                       dropout, activation, self.bond_ffn_base, constraint, ffn_type='bond', shared_ffn))
```

File: chemprop/models/ffn.py [hunk 3]
```diff
@@ -83,27 +114,36 @@ def __init__(self,
...
 
         if constraint:
-            self.ffn = DenseLayers(first_linear_dim=features_size,
-                                   hidden_size=hidden_size,
-                                   num_layers=num_layers,
-                                   output_size=hidden_size,
-                                   dropout=dropout,
-                                   activation=activation)
+            if shared_ffn:
+                self.ffn = ffn_base
```
Comment (shihchengli) [inline chemprop/models/ffn.py:137]:
```suggestion
                self.ffn = nn.Sequential([
                    ffn_base,
                    DenseLayers(
                        first_linear_dim=features_size,
                        hidden_size=hidden_size,
                        num_layers=1,
                        output_size=hidden_size,
                        dropout=dropout,
                        activation=activation
                    )
                ])
```

File: chemprop/models/ffn.py [hunk 4]
```diff
@@ -117,12 +157,27 @@ def __init__(self,
...
         else:
-            self.ffn_readout = DenseLayers(first_linear_dim=features_size,
-                                           hidden_size=hidden_size,
-                                           num_layers=num_layers,
-                                           output_size=output_size,
-                                           dropout=dropout,
-                                           activation=activation)
+            if shared_ffn:
+                self.ffn_readout = nn.Sequential([
+                    self.ffn,
```
Comment (shihchengli) [inline chemprop/models/ffn.py:162]:
```suggestion
                    ffn_base,
```

**Review by shihchengli — CHANGES_REQUESTED** [review]
Thank you Chas! That is very clear, only something needed to be minorly modified. In addition to these changes, could you add a description of this in the README.md? If you want to test this code, you could use the below command with files in tests/data/ folder:
```
Chemprop_train \
--data_path tests/data/atomic_bond_regression.csv \
--dataset_type regression \
--is_atom_bond_targets \
--constraints_path tests/data/atomic_bond_constraints.csv \
--adding_h
```

File: chemprop/models/ffn.py (hunk 3 above)
Comment (cjmcgill) [reply chemprop/models/ffn.py:137]: I don't understand this suggestion, why would we need another layer here? ffn_base as currently constituted uses all but the last layer of the ffn, which is then added as the readout layer. I think actually my problem might be that the else statement in L139 uses the full number of layers and not num_layers - 1.

**Comment by shihchengli** [comment]
Looks good to me! Thanks!

**Pull request merged**